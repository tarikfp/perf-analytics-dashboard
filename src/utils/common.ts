/* eslint-disable arrow-body-style */
import { PerformanceResourceTimingHandler } from '@/types';

// helpers

const convertToSec = (value: number): number => value / 1000;

const takeFirstDigits = (value: number): number => Number(value.toFixed(3));

const findDisplayNameFromKey = <T>(
  data: T,
  arr: Array<{ id: keyof T | string; display: string }>,
): string[] => {
  const keys: string[] = [];
  arr.forEach((x) => {
    if (Object.keys(data).includes(x.id as string)) {
      keys.push(x.display);
    }
  });
  return keys;
};

const parsePerformanceResourceTimingData = (
  data: PerformanceResourceTimingHandler[],
): PerformanceResourceTimingHandler[] => {
  return data.map((item) => ({
    ...item,
    responseEnd: takeFirstDigits(convertToSec(item.responseEnd)),
    duration: takeFirstDigits(convertToSec(item.duration)),
    startTime: takeFirstDigits(convertToSec(item.startTime)),
  }));
};

const sortByDate = <T>(data: T[], sortKey: keyof T) => {
  return data.sort((a, b) => {
    const dateA = new Date(Number(a[sortKey]));
    const dateB = new Date(Number(b[sortKey]));
    return dateB.getTime() - dateA.getTime();
  });
};

export {
  sortByDate,
  takeFirstDigits,
  findDisplayNameFromKey,
  parsePerformanceResourceTimingData,
};
