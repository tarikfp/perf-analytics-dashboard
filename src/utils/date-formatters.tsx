import * as dateFns from 'date-fns';

// date formatters

const formatAxisDate = (date: string): string =>
  dateFns.format(new Date(date), 'HH:mm:ss');

const formatParamDate = (date: Date): string =>
  dateFns.format(date, 'yyyy-MM-dd HH:mm');

const formatTableDate = (date: string): string =>
  dateFns.format(new Date(date), 'yyyy-MM-dd HH:mm:ss');

export { formatAxisDate, formatParamDate, formatTableDate };
