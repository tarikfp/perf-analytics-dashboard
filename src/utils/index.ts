import dayjs from 'dayjs';

const formatTooltipDate = (date: string): string =>
  dayjs(date).format('YYYY/MM/DD HH:mm:ss');

const formatAxisDate = (date: string): string => dayjs(date).format('HH:mm:ss');

export { formatAxisDate, formatTooltipDate };
