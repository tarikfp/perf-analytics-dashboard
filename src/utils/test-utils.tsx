import * as React from 'react';
import * as renderer from 'react-test-renderer';
import { ThemeContext } from '../theme';

const themeContextValue = {
  isDarkMode: false,
  toggleDarkMode: () => {},
};

// helpers to create test component with theme context

export function createWithThemeContext(
  children: React.ReactElement | React.ReactNode,
): renderer.ReactTestRendererJSON | renderer.ReactTestRendererJSON[] | null {
  return renderer
    .create(
      <ThemeContext.Provider value={themeContextValue}>
        {children}
      </ThemeContext.Provider>,
    )
    .toJSON();
}

export function wrapWithThemeContext(
  children: React.ReactElement | React.ReactNode,
): React.ReactElement {
  return (
    <ThemeContext.Provider value={themeContextValue}>
      {children}
    </ThemeContext.Provider>
  );
}
