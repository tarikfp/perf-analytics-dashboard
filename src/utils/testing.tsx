import { render, RenderResult } from '@testing-library/react';
import { ThemeContext } from '@theme';
import * as React from 'react';
import * as renderer from 'react-test-renderer';

const themeContextValue = {
  isDarkMode: false,
  toggleDarkMode: () => null,
};

// helpers to create test component with theme context

export function createWithThemeContext(
  children: React.ReactElement | React.ReactNode,
): renderer.ReactTestRendererJSON | renderer.ReactTestRendererJSON[] | null {
  return renderer
    .create(
      <ThemeContext.Provider value={themeContextValue}>
        {children}
      </ThemeContext.Provider>,
    )
    .toJSON();
}

export function renderWithThemeContext(
  children: React.ReactElement | React.ReactNode,
): RenderResult {
  return render(
    <ThemeContext.Provider value={themeContextValue}>
      {children}
    </ThemeContext.Provider>,
  );
}

export function wrapWithThemeContext(
  children: React.ReactElement | React.ReactNode,
): React.ReactElement {
  return (
    <ThemeContext.Provider value={themeContextValue}>
      {children}
    </ThemeContext.Provider>
  );
}
