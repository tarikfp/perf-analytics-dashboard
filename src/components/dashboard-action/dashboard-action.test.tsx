import { createWithThemeContext, renderWithThemeContext } from '@/utils';
import * as React from 'react';
import DashboardAction from './dashboard-action';

// Date pickers have been tested internally by the mui pickers community.
// @see https://github.com/mui/material-ui-pickers/blob/next/lib/src/__tests__/DatePickerTestingLib.test.tsx

const props = {
  startDate: new Date(),
  endDate: new Date(),
  onStartDateChange: jest.fn(),
  onEndDateChange: jest.fn(),
  onFetchBetweenDatesClick: jest.fn(),
  onFetchLatestClick: jest.fn(),
};

test('date pickers renders correctly', () => {
  const tree = createWithThemeContext(<DashboardAction {...props} />);
  expect(tree).toMatchSnapshot();
});

test('submit button renders correctly', () => {
  const { getByText } = renderWithThemeContext(<DashboardAction {...props} />);
  expect(getByText('Create metric data')).toBeTruthy();
  expect(getByText('Fetch latest (last 30 mins)')).toBeTruthy();
  expect(getByText('Fetch between dates')).toBeTruthy();
});
