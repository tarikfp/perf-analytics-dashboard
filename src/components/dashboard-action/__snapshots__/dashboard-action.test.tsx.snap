// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`date pickers renders correctly 1`] = `
<div
  className="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation3 css-3su884-MuiPaper-root"
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#ffffff",
      "borderRadius": 4,
      "justifyContent": "center",
      "marginTop": 20,
      "minHeight": 150,
      "padding": 10,
    }
  }
>
  <div
    className="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2 css-1pckxps-MuiGrid-root"
  >
    <div
      className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-6 MuiGrid-grid-md-3 css-1hi6wun-MuiGrid-root"
    >
      <div
        className="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
      >
        <label
          className="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
          data-shrink={false}
          htmlFor="mui-1"
          id="mui-1-label"
        >
          Start date
        </label>
        <div
          className="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
          onClick={[Function]}
        >
          <input
            aria-invalid={false}
            aria-label="Choose date, selected date is May 4, 2022"
            aria-readonly={true}
            autoFocus={false}
            className="MuiOutlinedInput-input MuiInputBase-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
            id="mui-1"
            onAnimationStart={[Function]}
            onBlur={[Function]}
            onChange={[Function]}
            onClick={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            readOnly={true}
            required={false}
            type="text"
            value="05/04/2022 01:57 pm"
          />
          <fieldset
            aria-hidden={true}
            className="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
          >
            <legend
              className="css-1ftyaf0"
            >
              <span>
                Start date
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
    </div>
    <div
      className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-6 MuiGrid-grid-md-3 css-10voxkt-MuiGrid-root"
    >
      <div
        className="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
      >
        <label
          className="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
          data-shrink={false}
          htmlFor="mui-3"
          id="mui-3-label"
        >
          End date
        </label>
        <div
          className="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
          onClick={[Function]}
        >
          <input
            aria-invalid={false}
            aria-label="Choose date, selected date is May 4, 2022"
            aria-readonly={true}
            autoFocus={false}
            className="MuiOutlinedInput-input MuiInputBase-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
            id="mui-3"
            onAnimationStart={[Function]}
            onBlur={[Function]}
            onChange={[Function]}
            onClick={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            readOnly={true}
            required={false}
            type="text"
            value="05/04/2022 01:57 pm"
          />
          <fieldset
            aria-hidden={true}
            className="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
          >
            <legend
              className="css-1ftyaf0"
            >
              <span>
                End date
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
    </div>
    <div
      className="MuiBox-root css-14ajw3y"
    >
      <div
        aria-label="small button group"
        className="MuiButtonGroup-root MuiButtonGroup-outlined css-iajp3t-MuiButtonGroup-root"
        role="group"
      >
        <button
          className="MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeSmall MuiButton-containedSizeSmall MuiButtonBase-root MuiButtonGroup-grouped MuiButtonGroup-groupedHorizontal MuiButtonGroup-groupedOutlined MuiButtonGroup-groupedOutlinedHorizontal MuiButtonGroup-groupedOutlinedPrimary MuiButtonGroup-grouped MuiButtonGroup-groupedHorizontal MuiButtonGroup-groupedOutlined MuiButtonGroup-groupedOutlinedHorizontal MuiButtonGroup-groupedOutlinedPrimary css-eu6c0d-MuiButtonBase-root-MuiButton-root"
          disabled={false}
          onBlur={[Function]}
          onClick={[MockFunction]}
          onContextMenu={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={0}
          type="button"
        >
          Fetch between dates
          <span
            className="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
        <button
          className="MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeSmall MuiButton-containedSizeSmall MuiButtonBase-root MuiButtonGroup-grouped MuiButtonGroup-groupedHorizontal MuiButtonGroup-groupedOutlined MuiButtonGroup-groupedOutlinedHorizontal MuiButtonGroup-groupedOutlinedPrimary MuiButtonGroup-grouped MuiButtonGroup-groupedHorizontal MuiButtonGroup-groupedOutlined MuiButtonGroup-groupedOutlinedHorizontal MuiButtonGroup-groupedOutlinedPrimary css-eu6c0d-MuiButtonBase-root-MuiButton-root"
          disabled={false}
          onBlur={[Function]}
          onClick={[MockFunction]}
          onContextMenu={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={0}
          type="button"
        >
          Fetch latest (last 30 mins)
          <span
            className="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
        <button
          className="MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeSmall MuiButton-containedSizeSmall MuiButtonBase-root MuiButtonGroup-grouped MuiButtonGroup-groupedHorizontal MuiButtonGroup-groupedOutlined MuiButtonGroup-groupedOutlinedHorizontal MuiButtonGroup-groupedOutlinedPrimary MuiButtonGroup-grouped MuiButtonGroup-groupedHorizontal MuiButtonGroup-groupedOutlined MuiButtonGroup-groupedOutlinedHorizontal MuiButtonGroup-groupedOutlinedPrimary css-eu6c0d-MuiButtonBase-root-MuiButton-root"
          disabled={false}
          onBlur={[Function]}
          onClick={[Function]}
          onContextMenu={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={0}
          type="button"
        >
          Create metric data
          <span
            className="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
  </div>
</div>
`;
